docs/proposals/20210305-sidecarset-hotupgrade.md

```diff
@@ -0,0 +1,214 @@
+User use SidecarSet to be easily injecting and upgrading sidecar containers, the current upgrade can only meet the requirements of stateless sidecar containers.
+A stateless container can be upgraded directly through container restart, without migration of state between the old and new versions, for example: Logtail.
+
+However, there are many scenarios for stateful sidecar containers, such as Istio Envoy, which cannot be upgraded by a direct restart.
+The envoy intercepts all the traffic from the pod, and if you restart it directly, it will affect the availability of the business container.
+
+In this proposal, we try to figure out a way to support lossless hot upgrades for stateful sidecar containers.
+
+## Proposal
+
+The SidecarSet implements a hot upgrade mechanism of the stateful sidecar container by adding an empty container. As shown below:
+
+```yaml
+apiVersion: apps.kruise.io/v1alpha1
+kind: SidecarSet
+metadata:
+  name: test-sidecarset
+spec:
+  containers:
+  - name: sidecar-name
+    ....
+    # sidecar container upgrade method
+    upgradeStrategy:
+      upgradeType: HotUpgrade | ColdUpgrade
+      hotUpgradeEmptyImage: {Empty Image}
+```
+- upgradeType: HotUpgrade indicates hot upgrade for stateful sidecar container; ColdUpgrade is used for stateless sidecar container
+- hotUpgradeEmptyImage: when upgradeType=HotUpgrade, user needs to provide an empty container for hot upgrades
+
+**Note: hotUpgradeEmptyImage has the same configuration as the sidecar container, for example: command, lifecycle, probe, etc, but it doesn't do anything.**
+
+Hot upgrade consists of the following two processes:
+- inject hot upgrade sidecar containers
+- in-place hot upgrade sidecar container
+
+### Inject Containers
+
+When the sidecar container upgradeStrategy=HotUpgrade, the SidecarSet Webhook will inject two containers into the Pod:
+1. {sidecarContainer.name}-1: as shown in the figure below: envoy-1, the container run the actual working sidecar container, such as envoy:1.16.0
+2. {sidecarContainer.name}-2: as shown in the figure below: envoy-2, the container run the hot upgrade empty container, and it doesn't have to deal with any real logic, as long as it stays in place, such as empty:1.0
+
+![sidecarset hotupgrade_injection](../../docs/img/sidecarset_hotupgrade_injection.png)
+
+### Hot Upgrade
+
+The SidecarSet Controller breaks down the hot upgrade process of the sidecar container into three steps:
+1. Upgrade: upgrade the empty container to the new version of the sidecar container, such as envoy-2.Image = envoy:1.17.0
+2. Migration: the process completes the state migration of stateful container, which needs to be done by the business itself according to its own container characteristics.
+PostStartHook completes the migration of the above process.
+(**Note: PostStartHook must be block during the migration, and exit when migration complete.**)
```

Comment:
PostStartHook must block during the migration

Reply:
In addition, you need to specify which container's PostStartHook you are referring to here.